README.md
# Requeuest

[![pipeline status][badge-pipeline-img]][badge-pipeline-url]
[![coverage report][badge-coverage-img]][badge-coverage-url]
[![documentation (docs.rs)][badge-docs-release-img]][badge-docs-release-url]
[![documentation (main)][badge-docs-main-img]][badge-docs-main-url]

[badge-pipeline-img]: https://gitlab.com/famedly/company/backend/libraries/requeuest/badges/main/pipeline.svg
[badge-pipeline-url]: https://gitlab.com/famedly/company/backend/libraries/requeuest/-/commits/main
[badge-coverage-img]: https://gitlab.com/famedly/company/backend/libraries/requeuest/badges/main/coverage.svg
[badge-coverage-url]: https://gitlab.com/famedly/company/backend/libraries/requeuest/-/commits/main
[badge-docs-release-img]: https://img.shields.io/crates/v/requeuest?label=docs&color=informational
[badge-docs-release-url]: https://docs.rs/requeuest
[badge-docs-main-img]: https://img.shields.io/badge/docs-main-informational
[badge-docs-main-url]: https://famedly.gitlab.io/company/backend/libraries/requeuest/requeuest/index.html

![requeuest logo](logo.svg)

Requeuest (pronounced "recused") is a message queue which acts as an intermediary for HTTP requests, making sure that the sent request gets successfully delivered eventually, meaning that you do not have to implement retry logic for HTTP API requests. The queue uses the [`sqlxmq`] crate to make postgres its store for messages, which avoids the reliability risk of a dedicated message queue service potentially being down. This comes with the trade-off that job runners become part of the library consumer's process, and that a handle to the runner has to be kept alive so jobs can run in the background, since jobs cannot be delegated to a separate runner service.

## Getting started
Assuming you already have an `sqlx` connection to a postgres database, you will first need to run migrations so the needed tables and SQL functions can get set up on your postgres database. It's recommended to have a distinct database specifically for requeuest, to avoid interference with the migration management of a different service or library.

```rust
requeuest::migrate(&pool).await?;
```

Once that's taken care of, start by constructing a client. This is what you will use to spawn requests, an what will execute jobs in the background. It will keep doing so until it is dropped. The client contains a tokio `JoinHandle` which you can remove from the client with the `Client::take_listener` method if you want the listener to keep running after the client has dropped, or otherwise interface with the background task directly.

```rust
use requeuest::Client;

let client = Client::new(pool, &["my_service"]).await?;
```

After the client has been constructed, you can begin spawning jobs. Here we send a get request to an example address:

```rust
use requeuest::{HeaderMap, Request};

let request = Request::get("https://foo.bar/_api/baz".parse()?, HeaderMap::new());
client.spawn("my_service", &request).await?;
```

You can also also get the response back from a successfully delivered request.

```rust
// You can skip the HeaderMap import by invoking the constructor via the Default trait
let request = Request::post("https://example.com/_api/bar/foo".parse()?, Vec::from("some data"), Default::default());
let response = client.spawn_returning("my_service", &request).await?;
```

Note that the `spawn_returning` method *will* wait indefinitely (or to be precise, roughly 10^293 years) until a successful response is received by default, so this will wait forever if a request is sent to e.g. an unregistered domain, or a request to an API which that's guaranteed to always get a response back with a non-200 response code.

[`sqlxmq`]: https://docs.rs/sqlxmq

## Testing
requeuest's integration tests rely on [`sqlx-database-tester`], which expects a postgres database to be available with a user that has permissions to create, edit and drop databases. Which database(s) to use is set with a postgres url (e.g. `postgres://postgres:password@localhost/`) in the `DATABASE_URL` environment variable, or with an entry in a `.env`. See `sqlx-database-tester`'s documentation for more information.

[`sqlx-database-tester`]: https://gitlab.com/famedly/company/backend/libraries/sqlx-database-tester

## Pre-commit usage

1. If not installed, install with your package manager, or `pip install --user pre-commit`
2. Run `pre-commit autoupdate` to update the pre-commit config to use the newest template
3. Run `pre-commit install` to install the pre-commit hooks to your local environment


CHANGELOG.md
# Changelog

## [0.6.0] - 2022-06-29

### Miscellaneous Tasks

- Upgrade sqlxmq to 0.4.0
- Upgrade sqlx to 0.6
- Upgrade uuid to 1.0

## [0.5.0] - 2022-06-10

### Miscellaneous Tasks

- Bump sqlxmq

## [0.4.0] - 2022-05-11

### Features

- Function to clear channels of pending jobs
- [**breaking**] Builder interface added to `Request`. Available via `Request::builder()`. ([!42])

### Changed

- The methods `Request::{get, head, delete, put, post}` now return a builder
  with the corresponding method. Additionally, they now expect only a url or a url and a body for `post` and `put`.  ([!42])

### Miscellaneous Tasks

- Move to upstream retry check
- Fix typos
- Add pre-commit
- Update files from template
- Update files from template

### Testing

- Check that AcceptedResponse matches correctly

[!42]: https://gitlab.com/famedly/company/backend/libraries/requeuest/-/merge_requests/42

## [0.3.0] - 2021-09-21

### Bug Fixes

- Fix constraint violation issues with ordered jobs ([!33])

### Removed

- Remove functions for spawning jobs via transactions ([!33]).

### Miscellaneous Tasks

- Correct url for documentation badge in readme ([!30])

[!30]: https://gitlab.com/famedly/company/backend/libraries/requeuest/-/merge_requests/30
[!33]: https://gitlab.com/famedly/company/backend/libraries/requeuest/-/merge_requests/33


## [0.2.1] - 2021-09-15

### Bug Fixes

- Fix configured request headers not being set on sent requests

### Documentation

- Document integration test setup in readme
- Misc readme improvements

### Internal

- Add missing metadata to Cargo.toml
- Add logo
- Add edition to rustfmt.toml
- Add .env to .gitignore

### Testing

- Add integration tests

### Ci

- Delegate further to CI template
- Add postgres service to the `check` job for the integration tests

## [0.2.0] - 2021-08-23

### Added

- Add conversion from request types from the `http` crate ([!20])

  Available via `Request::from_http_builder`, `Request::from_http_empty`, and `Request::from_http_body`.

### Changed

- Moved `Request::spawn_with` and related functions to `Client` ([!18])

### Internal

- Remove remaining global static, use `sqlxmq` context mechanism instead ([!18])
- Add .editorconfig ([!19])

[!18]: https://gitlab.com/famedly/company/backend/libraries/requeuest/-/merge_requests/18
[!19]: https://gitlab.com/famedly/company/backend/libraries/requeuest/-/merge_requests/19
[!20]: https://gitlab.com/famedly/company/backend/libraries/requeuest/-/merge_requests/20

## [0.1.1] - 2021-08-16
First release! 🎉


tests/send.rs
//! Tests that verify that HTTP requeuests are correctly sent

#![allow(clippy::unwrap_used, clippy::expect_used)]

use std::{
	iter::FromIterator,
	sync::atomic::{AtomicBool, AtomicU32, Ordering},
	time::Duration,
};

use requeuest::{
	self,
	client::{Channels, Client},
	request::Request,
	HeaderMap, Url,
};
use reqwest::header::{HeaderValue, AUTHORIZATION};
use tokio::sync::Notify;

static INSTALL_EYRE: std::sync::Once = std::sync::Once::new();

fn install_eyre() {
	INSTALL_EYRE.call_once(|| color_eyre::install().expect("Installing eyre failed"));
}

/// Crate a hyper `Service` with the given future as its `service_fn`
#[macro_export]
macro_rules! service {
	($closure:expr) => {
		hyper::service::make_service_fn(|_conn| async {
			Ok::<_, hyper::Error>(hyper::service::service_fn($closure))
		})
	};
}

/// Create a hyper `Server` with the given graceful shutdown closure, returning
/// the ip address and server object.
#[macro_export]
macro_rules! server {
	($service:expr, $shutdown:expr) => {{
		let addr = std::net::SocketAddr::from(([127, 0, 0, 1], 0));
		let server = hyper::Server::bind(&addr).serve($service);
		// Get the address from the server so we know what port was picked
		let addr = server.local_addr();
		let server = server.with_graceful_shutdown($shutdown);
		(addr, server)
	}};
}

static SEND_EMPTY_NOTIF: Notify = Notify::const_new();

/// Verifies that a request without a body is correctly sent
#[sqlx_database_tester::test(pool(variable = "pool", skip_migrations))]
#[ntest::timeout(60_000)]
async fn send_empty() -> color_eyre::eyre::Result<()> {
	install_eyre();
	requeuest::migrate(&pool).await?;
	let client = Client::new(pool, Channels::All).await?;

	let service = service!(|req| async move {
		assert_eq!(req.method(), hyper::Method::GET, "Wrong method");
		assert_eq!(req.uri().path(), "/path", "Wrong URI path");
		assert_eq!(req.uri().query().unwrap(), "query=foo&param=bar", "Wrong URI query");
		assert_eq!(req.headers()[AUTHORIZATION], &"Bearer: secret", "Wrong HTTP header");
		SEND_EMPTY_NOTIF.notify_one();
		Ok::<_, hyper::Error>(hyper::Response::new(hyper::Body::from("OK")))
	});

	let (addr, server) = server!(service, async { SEND_EMPTY_NOTIF.notified().await });

	let headers =
		HeaderMap::from_iter([(AUTHORIZATION, HeaderValue::from_static("Bearer: secret"))]);
	let request = Request::get(format!("http://{}/path?query=foo&param=bar", addr).as_str())?
		.headers(headers)
		.build();

	client.spawn("channel", &request).await?;

	server.await?;

	Ok(())
}

static RETRY_COUNT: AtomicU32 = AtomicU32::new(0);
static RETRY_NOTIF: Notify = Notify::const_new();

/// Verifies that a request is correctly retried
#[sqlx_database_tester::test(pool(variable = "pool", skip_migrations))]
#[ntest::timeout(60_000)]
async fn retry() -> color_eyre::eyre::Result<()> {
	install_eyre();
	requeuest::migrate(&pool).await?;
	let client = Client::new(pool, Channels::All).await?;

	let service = service!(|_req| async move {
		let attempt = RETRY_COUNT.fetch_add(1, Ordering::SeqCst);
		let response = match attempt {
			0..=2 => {
				hyper::Response::builder().status(400).body(hyper::Body::from("Try again")).unwrap()
			}
			3 => {
				RETRY_NOTIF.notify_one();
				hyper::Response::new(hyper::Body::from("OK"))
			}
			_ => panic!("Too many retries!"),
		};
		Ok::<_, hyper::Error>(response)
	});

	let (addr, server) = server!(service, async { RETRY_NOTIF.notified().await });

	let request = Request::get(format!("http://{}/", addr).as_str())?.build();

	client
		.spawn_cfg("channel", &request, |req| {
			req.set_retries(3);
			req.set_retry_backoff(Duration::from_millis(10));
		})
		.await?;

	server.await?;

	Ok(())
}

static ORDER_NOTIF: Notify = Notify::const_new();
static ORDER_REQ_NUM: AtomicU32 = AtomicU32::new(1);

/// Verifies that returning requests get delivered sequentially
#[sqlx_database_tester::test(pool(variable = "pool", skip_migrations))]
#[ntest::timeout(30_000)]
async fn order() -> color_eyre::eyre::Result<()> {
	install_eyre();
	requeuest::migrate(&pool).await?;
	let client = Client::new(pool, Channels::All).await?;

	let service = service!(|req: hyper::Request<hyper::Body>| async move {
		let num = ORDER_REQ_NUM.fetch_add(1, Ordering::AcqRel);
		let body = hyper::body::to_bytes(req.into_body()).await?;
		let req_num = String::from_utf8_lossy(&body).parse::<u32>();

		if req_num != Ok(num) {
			panic!("Wrong order");
		}
		if num == 3 {
			ORDER_NOTIF.notify_one();
		}

		Ok::<_, hyper::Error>(hyper::Response::new(hyper::Body::from("OK")))
	});

	let (addr, server) = server!(service, async { ORDER_NOTIF.notified().await });

	let url: Url = format!("http://{}/", addr).parse()?;
	let request1 = Request::post(url.clone(), b"1".to_vec())?.build();
	let request2 = Request::post(url.clone(), b"2".to_vec())?.build();
	let request3 = Request::post(url.clone(), b"3".to_vec())?.build();

	let handle = tokio::spawn(async move { server.await });

	let cfg = |job: &mut sqlxmq::JobBuilder| {
		job.set_ordered(true);
	};
	client.spawn_cfg("order", &request1, cfg).await?;
	client.spawn_cfg("order", &request2, cfg).await?;
	client.spawn_cfg("order", &request3, cfg).await?;

	handle.await??;

	Ok(())
}

static RECEIVED_REQUEST: AtomicBool = AtomicBool::new(false);

/// Verifies that clearing requests for a queue works
#[sqlx_database_tester::test(pool(variable = "pool", skip_migrations))]
#[ntest::timeout(30_000)]
async fn clear() -> color_eyre::eyre::Result<()> {
	install_eyre();
	requeuest::migrate(&pool).await?;
	let client = Client::new(pool, Channels::All).await?;

	let service = service!(|_| async move {
		RECEIVED_REQUEST.store(true, Ordering::SeqCst);
		Ok::<_, hyper::Error>(hyper::Response::new(hyper::Body::from("ERR")))
	});

	let (addr, server) =
		server!(service, async { tokio::time::sleep(Duration::from_secs(1)).await });

	let request = Request::get(format!("http://{}/", addr).as_str())?.build();

	client
		.spawn_cfg("clear", &request, |req| {
			req.set_delay(Duration::from_millis(500));
		})
		.await?;
	client.clear(Channels::List(&["clear"])).await?;

	server.await?;

	assert!(!RECEIVED_REQUEST.load(Ordering::SeqCst), "Failed to clear");

	Ok(())
}


src/job.rs
//! Contains the definition of the job which sends http requests.

use std::{
	collections::HashMap,
	sync::{Arc, LockResult, Mutex, MutexGuard},
};

use sqlxmq::{job, CurrentJob};
use tokio::sync::oneshot;
use uuid::Uuid;

use crate::{error::JobError, request::Request};

/// Alias for the result type sqlxmq jobs expect.
pub type JobResult = Result<(), Box<dyn std::error::Error + Send + Sync + 'static>>;

/// Alias for a map from request UUID to associated oneshot sender
type SenderMap = HashMap<Uuid, oneshot::Sender<reqwest::Response>>;

/// Mechanism for returning responses from successful jobs.
#[derive(Debug)]
pub(crate) struct ResponseSender(Arc<Mutex<SenderMap>>);

impl ResponseSender {
	/// Constructs a new response sender.
	pub fn new() -> ResponseSender {
		ResponseSender(Arc::new(Mutex::new(HashMap::new())))
	}

	/// Gets a lock to the mutex wrapping the map.
	pub fn lock(&self) -> LockResult<MutexGuard<SenderMap>> {
		self.0.lock()
	}
}

impl Clone for ResponseSender {
	fn clone(&self) -> Self {
		ResponseSender(Arc::clone(&self.0))
	}
}

/// The function which runs HTTP jobs and actually sends the requests.
#[job(name = "http")]
pub async fn http(mut job: CurrentJob, client: reqwest::Client) -> JobResult {
	// validate the job payload
	let payload = job.raw_bytes().ok_or(JobError::MissingRequest)?;
	let request: Request = bincode::deserialize(payload)?;

	// construct and send the request
	let mut builder = client.request(request.method, request.url).headers(request.headers);
	if let Some(body) = request.body {
		builder = builder.body(body);
	}
	let response = builder.send().await?;

	// complete the job if the response is in the accepted set
	if request.accept_responses.iter().any(|accepted| accepted.accepts(response.status())) {
		job.complete().await?;
	}

	Ok(())
}

/// Sends the response to the HTTP request back via a oneshot channel.
#[job(name = "http_response")]
pub async fn http_response(
	mut job: CurrentJob,
	client: reqwest::Client,
	sender: ResponseSender,
) -> JobResult {
	// validate the job payload
	let payload = job.raw_bytes().ok_or(JobError::MissingRequest)?;
	let request: Request = bincode::deserialize(payload)?;

	// construct and send the request
	let mut builder = client.request(request.method, request.url);
	if let Some(body) = request.body {
		builder = builder.body(body);
	}
	let response = builder.send().await?;

	// complete the job if the response is in the accepted set
	if request.accept_responses.iter().any(|accepted| accepted.accepts(response.status())) {
		job.complete().await?;

		#[allow(clippy::unwrap_used)] // We don't handle poisoning
		let sender = sender.0.lock().unwrap().remove(&job.id()).ok_or(JobError::MissingSender)?;
		sender.send(response).or(Err(JobError::MissingReceiver))?;
	}

	Ok(())
}


src/lib.rs
//! Requeuest is a library for queueing the sending of HTTP requests. It's built
//! with the [sqlxmq](https://docs.rs/sqlxmq) crate, which is a message queue that uses a postgres database
//! for storing messages.
//!
//! ## Getting started
//! Assuming you already have an `sqlx` connection to a postgres database, you
//! will first need to run migrations so the needed tables and SQL functions can
//! get set up on your postgres database.
//!
//! ```no_run
//! # async fn test(pool: &sqlx::Pool<sqlx::Postgres>) -> Result<(), sqlx::migrate::MigrateError> {
//! requeuest::migrate(&pool).await?;
//! # Ok(())
//! # }
//! ```
//!
//! Once that's taken care of, start by constructing a client. This is what you
//! will use to spawn requests, an what will execute jobs in the background. It
//! will keep doing so until it is dropped. The client contains a tokio
//! `JoinHandle` which you can remove from the client with
//! [`Client::take_listener`](crate::Client::take_listener) if you want the
//! listener to keep running after the client has dropped, or otherwise
//! interface with the background task directly.
//!
//! ```no_run
//! # async fn test(pool: sqlx::Pool<sqlx::Postgres>) -> Result<(), sqlx::Error> {
//! use requeuest::{Client, client::Channels};
//!
//! let client = Client::new(pool, Channels::List(&["my_service"])).await?;
//! # Ok(())
//! # }
//! ```
//!
//! After the client has been constructed, you can begin spawning jobs. Here we
//! send a get request to an example address:
//!
//! ```no_run
//! use requeuest::Request;
//!
//! # async fn test(client: requeuest::Client) -> Result<(), Box<dyn std::error::Error>> {
//! let request = Request::get("https://foo.bar/_api/baz")?.build();
//! client.spawn("my_service", &request).await?;
//! # Ok(())
//! # }
//! ```
//!
//! You can also also get the response back from a successfully delivered
//! request.
//!
//! ```no_run
//! # use requeuest::Request;
//!
//! # async fn test(client: requeuest::Client) -> Result<(), Box<dyn std::error::Error>> {
//! let request = Request::post("https://example.com/_api/bar/foo", Vec::from("some data"))?.build();
//! let response = client.spawn_returning("my_service", &request).await?;
//! # Ok(())
//! # }
//! ```
//!
//! Note that the `spawn_returning` method *will* wait indefinitely (or to be
//! precise, roughly 10^293 years) until a successful response is received, so
//! this will wait forever if a request is sent to e.g. an unregistered domain,
//! or sends data to an API which will always result in a non-200 response code.
//!
//! # Features
//! This crate has the following features:
//! * `http`: Enable conversion of requests from the [`http`] crate
//! * Async runtime and TLS implementation for [`sqlx`]:
//!     * Any of `runtime-{tokio,actix,async-std}-{rustls,native-tls}`

#![doc(
	html_logo_url = "https://gitlab.com/famedly/company/backend/libraries/requeuest/-/raw/main/logo.svg"
)]
#![deny(missing_docs)]

pub mod client;
pub mod error;
pub(crate) mod job;
pub mod request;

pub use client::Client;
pub use request::Request;
pub use reqwest::{self, header::HeaderMap, Method};
pub use sqlx::{Pool, Postgres};
pub use url::{ParseError, Url};
pub use uuid::Uuid;

/// Runs the SQL migrations this library needs.
pub async fn migrate(pool: &Pool<Postgres>) -> Result<(), sqlx::migrate::MigrateError> {
	sqlx::migrate!().run(pool).await
}


src/request.rs
//! Contains the definition of the request which gets (de)serialized and sent to
//! the database

use std::{collections::HashSet, convert::TryInto};

use reqwest::{header::HeaderMap, Method, StatusCode};
use serde::{Deserialize, Serialize};
use typed_builder::TypedBuilder;
use url::Url;

/// An HTTP request to be sent through the job queue.
#[derive(Serialize, Deserialize, Debug, TypedBuilder)]
#[must_use]
pub struct Request {
	/// The url to send the request to.
	pub url: Url,
	/// The body of the request.
	#[builder(default, setter(strip_option, into))]
	pub body: Option<Vec<u8>>,
	/// The HTTP method to connect with
	#[serde(with = "http_serde::method")]
	pub method: Method,
	/// The HTTP headers to set for the request.
	#[serde(with = "http_serde::header_map")]
	#[builder(default)]
	pub headers: HeaderMap,
	/// A set of HTTP response codes which won't cause a retry.
	#[serde(default = "default_accepted_responses")]
	#[builder(default=default_accepted_responses())]
	pub accept_responses: HashSet<AcceptedResponse>,
}

/// The kinds of categories of response codes which a response can accept
#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Serialize, Deserialize)]
pub enum AcceptedResponse {
	/// Accept information responses, i.e. code range 100-199
	Informational,
	/// Accept success responses, i.e. code range 200-299.
	Success,
	/// Accept redirection responses, i.e. code range 300-399.
	Redirection,
	/// Accept client-side error responses, i.e. code range 400-499.
	ClientError,
	/// Accept server-side error responses, i.e. code range 500-599.
	ServerError,
	/// Accept a single specific response code.
	Single(u16),
	/// Accept an inclusive range of responses.
	Range(u16, u16),
}

impl AcceptedResponse {
	/// Check whether this acceptance filter accepts the given status code.
	#[must_use]
	pub fn accepts(self, status: StatusCode) -> bool {
		match self {
			AcceptedResponse::Informational => status.is_informational(),
			AcceptedResponse::Success => status.is_success(),
			AcceptedResponse::Redirection => status.is_redirection(),
			AcceptedResponse::ClientError => status.is_client_error(),
			AcceptedResponse::ServerError => status.is_server_error(),
			AcceptedResponse::Single(code) => status.as_u16() == code,
			AcceptedResponse::Range(min, max) => status.as_u16() >= min && status.as_u16() <= max,
		}
	}
}

/// Returns the set of responses which are considered valid by default
fn default_accepted_responses() -> HashSet<AcceptedResponse> {
	IntoIterator::into_iter([AcceptedResponse::Success]).into_iter().collect()
}

/// Return builder type for methods with predefined method
type WithUrlAndMethodBuilder = RequestBuilder<((Url,), (), (Method,), (), ())>;
/// Return builder type for methods with predefined method and body
type WithUrlAndBodyAndMethodBuilder =
	RequestBuilder<((Url,), (Option<Vec<u8>>,), (Method,), (), ())>;

impl Request {
	/// Constructs a `GET` request builder.
	///
	/// # Example
	/// ```
	/// # use requeuest::Request;
	/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
	/// Request::get("https://example.com")?.build();
	/// # Ok(())
	/// # }
	/// ```
	pub fn get<T>(url: T) -> Result<WithUrlAndMethodBuilder, <T as TryInto<Url>>::Error>
	where
		T: TryInto<Url>,
	{
		Ok(Request::builder().method(Method::GET).url(url.try_into()?))
	}

	/// Constructs a `HEAD` request builder.
	pub fn head<T>(url: T) -> Result<WithUrlAndMethodBuilder, <T as TryInto<Url>>::Error>
	where
		T: TryInto<Url>,
	{
		Ok(Request::builder().method(Method::HEAD).url(url.try_into()?))
	}

	/// Constructs a `DELETE` request builder.
	pub fn delete<T>(url: T) -> Result<WithUrlAndMethodBuilder, <T as TryInto<Url>>::Error>
	where
		T: TryInto<Url>,
	{
		Ok(Request::builder().method(Method::DELETE).url(url.try_into()?))
	}

	/// Constructs a `POST` request builder with the given body
	pub fn post<T>(
		url: T,
		body: impl Into<Vec<u8>>,
	) -> Result<WithUrlAndBodyAndMethodBuilder, <T as TryInto<Url>>::Error>
	where
		T: TryInto<Url>,
	{
		Ok(Request::builder().method(Method::POST).url(url.try_into()?).body(body))
	}

	/// Constructs a `PUT` request builder with the given body
	pub fn put<T>(
		url: T,
		body: impl Into<Vec<u8>>,
	) -> Result<WithUrlAndBodyAndMethodBuilder, <T as TryInto<Url>>::Error>
	where
		T: TryInto<Url>,
	{
		Ok(Request::builder().method(Method::PUT).url(url.try_into()?).body(body))
	}

	/// Convert a reqwest request into a requeuest request.
	pub fn from_reqwest(mut foreign: reqwest::Request) -> Self {
		Self {
			url: foreign.url().clone(),
			body: foreign.body().and_then(reqwest::Body::as_bytes).map(<[_]>::to_vec),
			method: std::mem::take(foreign.method_mut()),
			headers: std::mem::take(foreign.headers_mut()),
			accept_responses: default_accepted_responses(),
		}
	}

	/// Constructs a request by converting a request builder from the `http`
	/// crate. Returns `None` if the uri or the method are missing from the
	/// builder
	#[cfg(feature = "http")]
	pub fn from_http_builder(
		foreign: http::request::Builder,
		body: Option<Vec<u8>>,
	) -> Result<Self, crate::error::ConvertError> {
		match body {
			Some(body) => Ok(Self::from_http_body(foreign.body(body)?)?),
			None => Ok(Self::from_http_empty(foreign.body(())?)?),
		}
	}

	#[cfg(feature = "http")]
	fn from_http_parts(parts: http::request::Parts) -> Result<Self, url::ParseError> {
		Ok(Self {
			url: Url::parse(&parts.uri.to_string())?,
			body: None,
			method: parts.method,
			headers: parts.headers,
			accept_responses: default_accepted_responses(),
		})
	}

	/// Convert a [`http::Request`] with a body into a requeuest request.
	#[cfg(feature = "http")]
	pub fn from_http_body<B: Into<Vec<u8>>>(
		foreign: http::Request<B>,
	) -> Result<Self, url::ParseError> {
		let (parts, body) = foreign.into_parts();
		let mut request = Self::from_http_parts(parts)?;
		request.body = Some(body.into());
		Ok(request)
	}

	/// Convert a [`http::Request`] without a body into a requeuest request
	#[cfg(feature = "http")]
	pub fn from_http_empty<B>(foreign: http::Request<B>) -> Result<Self, url::ParseError> {
		let (parts, _) = foreign.into_parts();
		Self::from_http_parts(parts)
	}
}

#[cfg(test)]
mod tests {
	#![allow(clippy::unwrap_used)]
	use reqwest::{
		header::{HeaderMap, HeaderValue, AUTHORIZATION},
		Method, StatusCode,
	};
	use url::ParseError;

	use super::Request;

	/// Convenience function to convert a u16 to status code and unwrap the
	/// result
	fn status_code(code: u16) -> StatusCode {
		StatusCode::from_u16(code).unwrap()
	}

	/// Checks that `AcceptedResponse` accepts and rejects the right status
	/// codes
	#[test]
	fn accepted_range() {
		for code in (100..1000).map(status_code) {
			use super::AcceptedResponse::*;
			let num = code.as_u16();
			assert_eq!(Informational.accepts(code), (100..200).contains(&num));
			assert_eq!(Success.accepts(code), (200..300).contains(&num));
			assert_eq!(Redirection.accepts(code), (300..400).contains(&num));
			assert_eq!(ClientError.accepts(code), (400..500).contains(&num));
			assert_eq!(ServerError.accepts(code), (500..600).contains(&num));
			assert_eq!(Range(423, 489).accepts(code), (423..=489).contains(&num));
			assert_eq!(Single(200).accepts(code), num == 200);
		}
	}

	#[test]
	fn serialization() {
		let request =
			Request::post("https://example.com/", b"Some cool data".to_vec()).unwrap().build();
		let serialized = bincode::serialize(&request).unwrap();
		let deserialized: Request = bincode::deserialize(&serialized).unwrap();

		assert_eq!(request.url, deserialized.url);
		assert_eq!(request.method, deserialized.method);
		assert_eq!(request.body, deserialized.body);
		assert_eq!(request.headers, deserialized.headers);
	}

	#[test]
	fn convert_reqwest() {
		let mut foreign = reqwest::Request::new(Method::POST, "https://foo.bar/".parse().unwrap());
		foreign.headers_mut().insert(AUTHORIZATION, HeaderValue::from_static("Secret"));
		*foreign.body_mut() = Some("body".into());
		let request = Request::from_reqwest(foreign);

		assert_eq!(request.url.to_string(), "https://foo.bar/", "URL mismatch");
		assert_eq!(request.method, Method::POST, "Method mismatch");
		assert_eq!(request.headers.get(AUTHORIZATION).unwrap(), &"Secret", "Header mismatch");
		assert_eq!(request.body.unwrap(), b"body", "Body mismatch");
	}

	#[test]
	fn test_constructors() {
		let get_request = Request::get("http://get.example").unwrap().build();

		assert_eq!(get_request.url.to_string(), "http://get.example/", "URL mismatch");
		assert_eq!(get_request.method, Method::GET, "Method mismatch");
		assert_eq!(get_request.headers, HeaderMap::default(), "Header mismatch");
		assert_eq!(get_request.body, None, "Body mismatch");

		let delete_request = Request::delete("https://delete.example").unwrap().build();

		assert_eq!(delete_request.url.to_string(), "https://delete.example/", "URL mismatch");
		assert_eq!(delete_request.method, Method::DELETE, "Method mismatch");
		assert_eq!(delete_request.headers, HeaderMap::default(), "Header mismatch");
		assert_eq!(delete_request.body, None, "Body mismatch");

		let head_request = Request::head("https://head.example").unwrap().build();

		assert_eq!(head_request.url.to_string(), "https://head.example/", "URL mismatch");
		assert_eq!(head_request.method, Method::HEAD, "Method mismatch");
		assert_eq!(head_request.headers, HeaderMap::default(), "Header mismatch");
		assert_eq!(head_request.body, None, "Body mismatch");

		let post_request =
			Request::post("https://post.example", b"example post".to_vec()).unwrap().build();

		assert_eq!(post_request.url.to_string(), "https://post.example/", "URL mismatch");
		assert_eq!(post_request.method, Method::POST, "Method mismatch");
		assert_eq!(post_request.headers, HeaderMap::default(), "Header mismatch");
		assert_eq!(post_request.body.unwrap(), b"example post", "Body mismatch");

		let put_request =
			Request::put("https://put.example", b"example put".to_vec()).unwrap().build();

		assert_eq!(put_request.url.to_string(), "https://put.example/", "URL mismatch");
		assert_eq!(put_request.method, Method::PUT, "Method mismatch");
		assert_eq!(put_request.headers, HeaderMap::default(), "Header mismatch");
		assert_eq!(put_request.body.unwrap(), b"example put", "Body mismatch");
	}

	#[test]
	fn test_builder() {
		let mut header_map = HeaderMap::new();
		header_map.insert("AUTHORIZATION", "secret".parse().unwrap());
		let request = Request::builder()
			.method(Method::GET)
			.url("https://foo.bar/".parse().unwrap())
			.headers(header_map.clone())
			.build();

		assert_eq!(request.url.to_string(), "https://foo.bar/", "URL mismatch");
		assert_eq!(request.method, Method::GET, "Method mismatch");
		assert_eq!(request.headers, header_map, "Header mismatch");
		assert_eq!(request.body, None, "Body mismatch");
		assert_eq!(request.accept_responses, crate::request::default_accepted_responses());

		let request = Request::builder()
			.url("https://foo.bar/".parse().unwrap())
			.method(Method::POST)
			.body("body")
			.build();

		assert_eq!(request.url.to_string(), "https://foo.bar/", "URL mismatch");
		assert_eq!(request.method, Method::POST, "Method mismatch");
		assert_eq!(request.headers, HeaderMap::new(), "Header mismatch");
		assert_eq!(request.body.unwrap(), b"body", "Body mismatch");
	}

	#[test]
	fn test_url_parse_error() {
		let parse_error = Request::delete("test.de").err().unwrap();

		assert_eq!(parse_error, ParseError::RelativeUrlWithoutBase, "Error missmatch");
	}

	#[cfg(feature = "http")]
	#[test]
	fn convert_http_builder() {
		let foreign = http::Request::post("https://foo.bar/").header(AUTHORIZATION, "Secret");
		let request = Request::from_http_builder(foreign, Some(b"body".to_vec())).unwrap();

		assert_eq!(request.url.to_string(), "https://foo.bar/", "URL mismatch");
		assert_eq!(request.method, Method::POST, "Method mismatch");
		assert_eq!(request.headers.get(AUTHORIZATION).unwrap(), &"Secret", "Header mismatch");
		assert_eq!(request.body.unwrap(), b"body", "Body mismatch");

		let bad = http::request::Builder::new();
		assert!(Request::from_http_builder(bad, None).is_err(), "Missing value guard failed");
	}

	#[cfg(feature = "http")]
	#[test]
	fn convert_http_empty() {
		let foreign = http::Request::put("https://bar.baz/")
			.header(AUTHORIZATION, "Credentials")
			.body(())
			.unwrap();
		let request = Request::from_http_empty(foreign).unwrap();

		assert_eq!(request.url.to_string(), "https://bar.baz/", "URL mismatch");
		assert_eq!(request.method, Method::PUT, "Method mismatch");
		assert_eq!(request.headers.get(AUTHORIZATION).unwrap(), &"Credentials", "Header mismatch");
		assert_eq!(request.body, None, "Body unexpectedly not empty");
	}

	#[cfg(feature = "http")]
	#[test]
	fn convert_http_body() {
		let foreign = http::Request::delete("http://web.site/thing")
			.header(AUTHORIZATION, "Bearer: l0tsofl3tters")
			.body("yeet the thing")
			.unwrap();
		let request = Request::from_http_body(foreign).unwrap();

		assert_eq!(request.url.to_string(), "http://web.site/thing", "URL mismatch");
		assert_eq!(request.method, Method::DELETE, "Method mismatch");
		assert_eq!(
			request.headers.get(AUTHORIZATION).unwrap(),
			&"Bearer: l0tsofl3tters",
			"Header mismatch"
		);
		assert_eq!(request.body.unwrap(), b"yeet the thing", "Body mismatch");
	}
}
